src/IO/H5/CombineH5.cpp

```diff
@@ -149,4 +157,186 @@ void combine_h5_vol(const std::vector<std::string>& file_names,
+  // error, but that's less versatile when using globs
+  if (h5_files_to_combine.size() == 1) {
+    file_system::copy(h5_files_to_combine[0], output_h5_filename);
+    if (verbosity >= Verbosity::Quiet) {
+      Parallel::printf("Done!\n");
+    }
+    return;
+  }
+
+  // For each dat subfile, this holds the number of times from each of the H5
+  // files to combine so that we always use the latest times. A nullopt means no
+  // times from that H5 file will be used. The std::vector should be the same
+  // length as the number of H5 files to combine
+  std::unordered_map<std::string, std::vector<std::optional<size_t>>>
+      num_time_map{};
+
+  // The outer loop is over dat files because we don't require different dat
+  // files to have the same times
+  for (const std::string& dat_filename : subfile_dat_names) {
+    num_time_map[dat_filename];
+    num_time_map.at(dat_filename).resize(h5_files_to_combine.size());
+
+    // The legend and version are sanity checks
+    std::optional<std::vector<std::string>> legend{};
+    std::optional<uint32_t> version{};
+    // Nullopt just means the first file we are looping over
+    std::optional<double> earliest_time{};
+    // We loop backwards to always ensure the "latest" time is used.
+    for (int i = static_cast<int>(h5_files_to_combine.size()) - 1; i >= 0;
+         i--) {
+      const auto index = static_cast<size_t>(i);
+      const std::string& filename = h5_files_to_combine[index];
+      const h5::H5File<h5::AccessType::ReadOnly> file_to_combine{filename};
+      const auto& dat_file = file_to_combine.get<h5::Dat>(dat_filename);
+      const auto dimensions = dat_file.get_dimensions();
+      if (not legend.has_value()) {
+        legend = dat_file.get_legend();
+        version = dat_file.get_version();
+      }
+
+      // Only grab the times for now
+      const Matrix times = dat_file.get_data_subset({0}, 0, dimensions[0]);
+
+      // Makes things easier below.
+      if (UNLIKELY(times.rows() == 0)) {
+        ERROR_NO_TRACE("No times in dat file " << dat_filename << " in H5 file "
+                                               << filename);
+      }
+      if (UNLIKELY(legend.value() != dat_file.get_legend())) {
+        ERROR_NO_TRACE("Legend of dat file "
+                       << dat_filename << " in H5 file " << filename
+                       << " doesn't match other H5 files.");
+      }
+      if (UNLIKELY(version.value() != dat_file.get_version())) {
+        ERROR_NO_TRACE("Version of dat file "
+                       << dat_filename << " in H5 file " << filename
+                       << " doesn't match other H5 files.");
+      }
+
+      // This is the first file. We can't make any decisions here so just store
+      // the number of times and the earliest time of this file
+      if (not earliest_time.has_value()) {
+        num_time_map.at(dat_filename)[index] = dimensions[0];
+        earliest_time = times(0, 0);
```

Comment:
This is not guarantee to be the earliest time. It's the first time written, but that doesn't have to be the earliest time, I think. I believe our observers do not enforce ordering on write.